system/lib/libc/emscripten_time.c

```diff
@@ -17,7 +17,7 @@
 __attribute__((__weak__)) long  timezone = 0;
 __attribute__((__weak__)) int   daylight = 0;
 __attribute__((__weak__)) char *tzname[2] = { 0, 0 };
-thread_local char buffer[128];
+thread_local char timezone_buffer[128];
```

Comment:
This can do inside the `emscripten_get_timezone` function.

Reply:
```
/root/cache/sysroot/include/threads.h:12:22: note: expanded from macro 'thread_local'
#define thread_local _Thread_local
                     ^
/root/project/system/lib/libc/emscripten_time.c:33:10: error: address of stack memory associated with local variable 'buffer' returned [-Werror,-Wreturn-stack-address]
  return buffer;
         ^~~~~~
2 errors generated.
```

Reply:
Maybe you also need `static` then?

Reply:
My understanding was that thread_local implied static, so I'm surprised that you would need it: https://stackoverflow.com/questions/22794382/are-c11-thread-local-variables-automatically-static#:~:text=Thread%20local%20storage%20is%20static,one%20instance%20of%20the%20variable.

Reply:
i think it does imply static, but the error is related to it being a "stack" variable which seems to be a little off, is there a reason you prefer static but local function variable vs just thread_local as global?

Reply:
Its its static that how can it be on the "stack"?  That doesn't make much sense.  Static (and thread_local) variables are by definition *not* on the stack.

I just like it better because of scoping.   I guess its ok if its file-scoped + static (so not visible outside the file), but honestly it makes more sense to be function-scoped.  I'm struggling to understand that error message.

Reply:
I guess thread_local does not imply static, since  the real here is:

```
emscripten_time.c:31:3: error: '_Thread_local' variables must have global storage
  thread_local char buffer[128];
```

Using `static thread_local` fixes it!

Reply:
@sbc100 What compiler and version is CI using to compile this? I am trying to reproduce in godbolt `thread_local` seesm to work fine with -Wall -Werror on clang 13

Reply:
@sbc100 thread_local is c++ only  emscripten_time is C

Reply:
Its emcc doing the compiling, which uses clang-15 (AKA tip of tree).   You should able to reproduce locally using emscripten.

Reply:
Ah, it look like in C++ thread_local implies static, but not in C.

Reply:
_Thread_local in C which is what musl defines thread_local to must be global 

```
source>:9:5: error: '_Thread_local' variables must have global storage
    _Thread_local char buffer[128];
```

Reply:
threads.h
```
#ifdef __cplusplus
extern "C" {
typedef unsigned long thrd_t;
#else
typedef struct __pthread *thrd_t;
#define thread_local _Thread_local
#endif
```

Reply:
@sbc100 How do you want me to proceed now?

Reply:
Since this is C code you must explicitly add the `static` specifier.   So use `static thread_local`

Reply:
To be clear the error might be a little misleading: `variables must have global storage`.   Here "global storage" also includes function-local statics.

Reply:
ahh makes sense. Thanks @sbc100